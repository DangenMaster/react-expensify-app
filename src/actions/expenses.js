// Action generators
export const addExpense = (payload = {}) => ({
  type: 'ADD_EXPENSE',
  payload
});

export const removeExpense = (payload = {}) => ({
  type: 'REMOVE_EXPENSE',
  payload
});

export const editExpense = (id, payload = {}) => ({
  type: 'EDIT_EXPENSE',
  id,
  payload
});