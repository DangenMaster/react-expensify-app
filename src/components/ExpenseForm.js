import React from "react";
import uuid from "uuid";
import moment from "moment";
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      id: props.expense ? props.expense.id : uuid(),
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment() ,
      calendarFocused: false,
      error: ''
    }; 
  }
  
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onCalendarFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount.' }));
    } else {
      this.setState(() => ({ error: '' }));
      const { id, description, note, amount, createdAt } = this.state;
      this.props.onSubmit({
        id,
        description, 
        note, 
        amount: parseFloat(amount, 10) * 100, 
        createdAt: createdAt.valueOf() 
      });
    }
  };
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input 
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange} />
          <input 
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange} />
          <SingleDatePicker
            date={this.state.createdAt} // momentPropTypes.momentObj or null
            onDateChange={this.onDateChange} // PropTypes.func.isRequired
            focused={this.state.calendarFocused} // PropTypes.bool
            onFocusChange={this.onCalendarFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false} />
          <textarea 
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange} >
          </textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;