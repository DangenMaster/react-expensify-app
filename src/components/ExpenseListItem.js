import React from "react";
import { Link } from "react-router-dom";

const ExpenseListItem = ({ amount, createdAt, description, id }) => (
  <div>
    <h3>
      <Link to={`/edit/${id}`}>
        {description}
      </Link>
    </h3>
    <p>{amount} - {createdAt}</p>
  </div>
);

export default ExpenseListItem;