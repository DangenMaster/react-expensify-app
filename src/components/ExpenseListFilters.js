import React from "react";
import { connect } from "react-redux";
import 'react-dates/initialize';
import { DateRangePicker } from "react-dates";
import { setTextFilter, sortByFilter, setStartDateFilter, setEndDateFilter } from "./../actions/filters";

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDateFilter(startDate);
    this.props.setEndDateFilter(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSelectChange = (e) => {
    e.preventDefault();
    this.props.sortByFilter(e.target.value);
  }
  render() {
    return(
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              type="text"
              className="text-input"
              placeholder="Search expenses"
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.onSelectChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              startDateId="expenseStartDateFilterId"
              endDate={this.props.filters.endDate}
              endDateId="expenseEndDateFilterId"
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setStartDateFilter: (startDate) => dispatch(setStartDateFilter({ startDate })),
  setEndDateFilter: (endDate) => dispatch(setEndDateFilter({ endDate })),
  setTextFilter: (text) => dispatch(setTextFilter({ text })),
  sortByFilter: (sortBy) => dispatch(sortByFilter({ sortBy }))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
