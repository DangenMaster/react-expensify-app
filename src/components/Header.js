import React from "react";
import { connect } from "react-redux";
import { NavLink } from "react-router-dom";
import { startGoogleLogout } from "../actions/auth";

export const Header = ({ startLogout }) => (
  <header>
    <h1>Expensify</h1>
    <NavLink exact to="/dashboard" activeClassName="is-active">Dashboard</NavLink>
    <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
    <button 
      onClick={startLogout}
    > Logout
    </button>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startGoogleLogout())
});

export default connect(null, mapDispatchToProps)(Header);