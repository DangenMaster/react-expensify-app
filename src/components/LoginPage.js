import React from "react";
import { connect } from "react-redux";
import { startGoogleLogin } from "../actions/auth";

export const LoginPage = ({ startLogin }) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">Expensify</h1>
      <p>It's time to get your expenses under control.</p>
      <button 
        className="button"
        onClick={startLogin}
      > 
        Login with Google 
      </button>
    </div>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startGoogleLogin())
});

export default connect(null, mapDispatchToProps)(LoginPage);