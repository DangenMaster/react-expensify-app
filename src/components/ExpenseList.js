import React from "react";
import { connect } from "react-redux";
import selectExpenses from "./../selectors/expenses";
import ExpenseListItem from "./ExpenseListItem";

export const ExpenseList = (props) => (
  <div>
  {
    props.expenses.length === 0 ? (
      <p>No Expenses</p>
    ) : (
      props.expenses.map((expense) => <ExpenseListItem key={expense.id} {...expense} />)
    )
  }
  </div>
);

const mapStateToProps = (state) => ({
  expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);