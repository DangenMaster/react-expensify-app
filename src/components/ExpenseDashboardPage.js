import React from "react";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpenseList from "./ExpenseList";
import ExpensesSummary from "./ExpensesSummary";

const ExpenseDashboardPage = () => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;