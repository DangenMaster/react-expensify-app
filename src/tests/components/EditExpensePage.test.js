import React from "react";
import { shallow } from "enzyme";
import expenses from "../fixtures/expenses";
import { EditExpensePage } from "../../components/EditExpensePage";

test('Should render EditExpensePage correctly', () => {
  const wrapper = shallow(<EditExpensePage />);
  expect(wrapper).toMatchSnapshot();
});

test('should handle edit expense', () => {
  const onSubmitSpy = jest.fn();
  const history = { push: jest.fn() };
  const wrapper = shallow(<EditExpensePage expense={expenses[0]} onSubmit={onSubmitSpy} history={history} />);
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
  expect(onSubmitSpy).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith('/');
});

test('should handle remove expense', () => {
  const onClickSpy = jest.fn();
  const history = { push: jest.fn() };
  const wrapper = shallow(<EditExpensePage expense={expenses[0]} onClick={onClickSpy} history={history} />);
  wrapper.find('button').simulate('click', { preventDefault: () => {} });
  expect(onClickSpy).toHaveBeenLastCalledWith(expenses[0].id);
  expect(history.push).toHaveBeenLastCalledWith('/');
});
