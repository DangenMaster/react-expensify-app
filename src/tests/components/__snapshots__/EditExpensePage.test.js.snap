// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
